drivers/adc/adc_stm32.c

```diff
@@ -1240,77 +1117,59 @@ static void adc_stm32_setup_speed(const struct device *dev, uint8_t id,
 		(const struct adc_stm32_cfg *)dev->config;
 	ADC_TypeDef *adc = config->base;
 
-#if defined(CONFIG_SOC_SERIES_STM32F0X) || defined(CONFIG_SOC_SERIES_STM32L0X)
-	LL_ADC_SetSamplingTimeCommonChannels(adc,
-		table_samp_time[acq_time_index]);
-#elif defined(CONFIG_SOC_SERIES_STM32C0X)
-	LL_ADC_SetSamplingTimeCommonChannels(adc, LL_ADC_SAMPLINGTIME_COMMON_1,
-					     table_samp_time[acq_time_index]);
-#elif defined(CONFIG_SOC_SERIES_STM32G0X)
-	/* Errata ES0418 and more: ADC sampling time might be one cycle longer */
-	if (acq_time_index  < 2) {
-		acq_time_index = 2;
-	}
-	LL_ADC_SetSamplingTimeCommonChannels(adc, LL_ADC_SAMPLINGTIME_COMMON_1,
-		table_samp_time[acq_time_index]);
-#elif defined(CONFIG_SOC_SERIES_STM32WLX)
-	LL_ADC_SetChannelSamplingTime(adc,
-				      __LL_ADC_DECIMAL_NB_TO_CHANNEL(id),
-				      LL_ADC_SAMPLINGTIME_COMMON_1);
-	LL_ADC_SetSamplingTimeCommonChannels(adc,
-					     __LL_ADC_DECIMAL_NB_TO_CHANNEL(id),
-					     table_samp_time[acq_time_index]);
-#elif defined(CONFIG_SOC_SERIES_STM32U5X)
-	if (adc != ADC4) {
+	if (config->sampling_time_common_channel_nb == 0) {
+#if !defined(CONFIG_SOC_SERIES_STM32F0X) && !defined(CONFIG_SOC_SERIES_STM32L0X)
 		LL_ADC_SetChannelSamplingTime(adc,
-					      __LL_ADC_DECIMAL_NB_TO_CHANNEL(id),
-					      table_samp_time[acq_time_index]);
+			__LL_ADC_DECIMAL_NB_TO_CHANNEL(id),
+			config->sampling_time_table[acq_time_index]);
+#endif
 	} else {
+#ifdef LL_ADC_SAMPLINGTIME_COMMON_1
+		LL_ADC_SetChannelSamplingTime(adc,
+			__LL_ADC_DECIMAL_NB_TO_CHANNEL(id),
+			LL_ADC_SAMPLINGTIME_COMMON_1);
 		LL_ADC_SetSamplingTimeCommonChannels(adc,
-						     LL_ADC_SAMPLINGTIME_COMMON_1,
-				       table_samp_time[acq_time_index]);
-	}
-#else
-	LL_ADC_SetChannelSamplingTime(adc,
-		__LL_ADC_DECIMAL_NB_TO_CHANNEL(id),
-		table_samp_time[acq_time_index]);
+			LL_ADC_SAMPLINGTIME_COMMON_1,
+			config->sampling_time_table[acq_time_index]);
+#elif defined(CONFIG_SOC_SERIES_STM32F0X) || defined(CONFIG_SOC_SERIES_STM32L0X)
+		LL_ADC_SetSamplingTimeCommonChannels(adc,
+			config->sampling_time_table[acq_time_index]);
 #endif
+	}
 }
```

Comment:
Do I understand this correctly?

The first `#if` here, in the true branch of the if statement, is to prevent a linker/compiler error because the F0/L0 don't provide `LL_ADC_SetChannelSamplingTime`.  This branch of the if statement should never occur on the F0/L0.  It could be written as:
```
if (config->sampling_time_common_channel_nb == 0) {
    #if defined(CONFIG_SOC_SERIES_STM32F0X) || defined(CONFIG_SOC_SERIES_STM32L0X)
    assert(false, "your device tree is broken");
    #else
    LL_ADC_SetChannelSamplingTime(adc, …);
    #endif
} else {
```

The 3rd `#if`, in the false branch, is for the same reason.  This is the `sampling_time_common_channel_nb == 1` branch, but without the `#if` it cause a compiler error on the non-F/L0X SoCs that don't have a two parameter `LL_ADC_SetSamplingTimeCommonChannels`.  Even though those SoC would never take this branch.

The false side could be written as:
```
} else {
    #if defined(CONFIG_SOC_SERIES_STM32F0X) || defined(CONFIG_SOC_SERIES_STM32L0X)
    assert(config->sampling_time_common_channel_nb == 1, "dts is wrong");
    LL_ADC_SetSamplingTimeCommonChannels(adc, ...);
   #else
    assert(config->sampling_time_common_channel_nb > 1, "dts is wrong");
    LL_ADC_SetChannelSamplingTime(adc, ..., COMMON_1);
    LL_ADC_SetSamplingTimeCommonChannels(adc, COMMON_1, ...);
   #endif
}
```

Reply:
You understand that correctly, the #if are only here to prevent compilation errors.
Your proposed modifications are perfectly correct, but I'm wondering if they would be actually useful.
The `sampling_time_common_channel_nb` property is defined in the dtsi, and is not supposed to be modified further by a user as it strictly describes the hardware. In that sense, it adds useless code.
But I also understand that it adds a check in case an error is made at some point.
I'd like a second opinion on this before adding or not. @erwango?

Reply:
I understand @xyzzy42 concern about having code branches that don't apply to some series which could feel awkward when reading but this is a compromise to make code a bit more digest, and the compiler will anyway do whatever optimizations are required.
Also, I don't think we should add too much code to validate dt entries. As mentioned by @gautierg-st these dts prop are descriptive and not supposed to be modified, they are supposedly correct.
We have some code dedicated to dt entries validation in some places, but this is rather done to prop that are supposed to be tuned by user.

Reply:
The example code I posted was just to be clear on how it worked, it wasn't meant to actually be used.  So, I certainly wouldn't use it.  I think if compile time checks can be done it can be useful to detect mistakes in the device tree.  Better yet is when the yaml can detect an invalid dts and produce a useful error message.  Runtime checks shouldn't be needed.

Here is something that is perhaps a bit better.  The basic idea is you can check with a macro if _any_ enabled ST ADC device(s) use a given `sampling_time_common_channel_nb` value.  So this can become something like:
```
if (config->sampling_time_common_channel_nb == 0) {
#if ANY_COMMON_SAMPLING_TIME_NB_IS(0)
        LL_ADC_SetChannelSamplingTime(adc, …);
#endif
} else if (config->sampling_time_common_channel_nb == 1) {
#if ANY_COMMON_SAMPLING_TIME_NB_IS(1)
        LL_ADC_SetSamplingTimeCommonChannels(adc, ...);
#endif
} else if (config->sampling_time_common_channel_nb == 2) {
#if ANY_COMMON_SAMPLING_TIME_NB_IS(2)
        LL_ADC_SetChannelSamplingTime(adc, ..., COMMON_1);
        LL_ADC_SetSamplingTimeCommonChannels(adc, COMMON_1, ...);
#endif
}
```

Advantage is the code used is controlled entirely from the device tree properties.  No list of STM32_F0X and so on.   And IMHO, I think maybe it's a bit easier to follow the code.

The macro is entirely possible and not that complex.

Reply:
Done